### DISCUSSION TITLE
Single socket

### DISCUSSION THREAD

**Discussion by artgodwin** [ole00/afterburner#25] [Ideas]
When building this circuit up, I started to wonder if I could use just a single socket instead of at least 3 for different device types.

I decided not to do it immediately as I could only find one type of GAL in my stock (I thought I had some 22V10s too) but it doesn't look very difficult. Probably a transistor to give a good ground sink to pin 10 when 16V8 is present, a few pulldowns on inputs that are also RA pins, and drives for pins 11,13 and 14, and a lookup table in the code.  That's with the GALs registered towards the top of the socket - it's a simpler translation with them at the bottom, but needs switching for EDIT supply between two pins. This could be a mechanical switch, though.

Most of the arduino's analog pins can also be used as digital out so I think there are just enough pins to do it. If a 32U4 is used the UART pins D0 and D1 are also freed up. I particularly like the Pro Micro (which is not an Arduino product) for this sort of thing, though they've got a lot more expensive recently.

**Comment by ole00** [comment]
These are very good ideas. Optionally an adapter for 16V8 could be used in the same way it is used for 20V8 (although an adapter is a worse solution compared to your idea IMO). I thought about other improvements as well (controlling the VPP programatically via digitally controllable resistor on I2C bus), but that would make the project a bit more complicated, possibly discouraging beginners to make it. Still, could be worth it - would you like to sketch a schematics for your new solution?

Analogue pins A0 - A5: spot on. When I created afterburner I did not know that they could be used as a regular digital GPIO, and I discovered that just recently.   So, yes this opens up more possibilities for improvements.

**Comment by artgodwin** [comment]
The schematic for the elm programmer http://elm-chan.org/works/pgal/pgal_e.png shows the commonised pin layout very well.

**Reply by ole00** [reply]
Thanks for that. I was not aware of pgal project.

**Comment by artgodwin** [comment]
Amazingly, both the control code and the fusemap assembler are written in x86 assembler. I can read it, but I'd rather not :).

**Comment by artgodwin** [comment]
Although that pgal circuit registers all the gals towards the bottom of the socket, and that seems sensible because so many pins are similar there, it means it has to switch Vpp and Vcc to pins at the top of the socket. Placing the GAL at the top of the socket so Vcc and Edit are shared makes the other pins less similar .. but they're almost all just logic levels so are easy to steer. Gnd on the bottom corner needs a strong driver, but that could be actual ground on the 22v10/20v8 and a transistor linking the 16v8 to gnd when required.

It's probably worth putting the 4k7 pulldowns everywhere, unless the AVR has a pulldown - I can't remember. 

It would be nice to use the 3608 even though it behaves badly at startup. It's very common and cheap. But the enable pin is awkward to get at. Maybe instead, remove the 2k2 resistor that acts as the bottom of the output voltage divider and let the FB pin go to output voltage. This should set it to 0.6V which is as good as off. Then have a couple of alternate resistors to switch in to give 12V and the adjustable value. Better still the digital pot. The only thing I've got against that is that I have every other part in my junkbox and only the pots would have to be ordered :). I don't know how typical my junkbox contents are. I do tend to have quite a few boost modules which are really the only special part in your existing design.

**Comment by ole00** [comment]
It is definitely tempting to throw in some new features to Afterburner, especially if they seem relatively easy to implement. But as we both agreed this extra complexity would probably discourage some beginners to build Afterburner in the first place. I would rather see the ATF16V8**C** supported and resolve the VCC stability during switch on. But maybe the only good solution for the VCC stability issue is to rework the circuit and keep the MT3608 module powered-on right from the start...

**Comment by artgodwin** [comment]
It might be worth checking the XL6009 converters. They also have to be modded for the enable pin but it may not be as fiddly. I don't know if they behave better but I think Ive got one or two to try. I also found a module on ebay with the enable pin brought out (not sure which converter). But they seem quite rare, probably not good to rely on them. The 6009 is as common as the  3608, but quite a lot bulkier.

**Comment by ole00** [comment]
What about removing the physical switch from the VCC side and adding it just between the ZIF GND pin and  GND. Actually use 2 switches - one physical and one transistor in series that would be switched by Arduino just before and after interaction with the GAL. The MT3608 would stay always on. Arduino software would have to be modified to strictly switch all pins into Inputs (Z) during inactivity to prevent GAL accidents.

**Reply by ole00** [reply]
On the second thought only the transistor based switch between ZIF GND pin and GND should be enough.  The physical switch might be better to use between the resistor arrays and GND to further isolate the GAL.

**Comment by ole00** [comment]
I checked XL6009 modules and they have even bigger caps and inductor values than MT3608. I suspect the voltage drop and overshoot - if pluged into current Afterburner design with the VCC switch - would cause even bigger voltage drop and overshoot. It's worth to try though.

**Comment by artgodwin** [comment]
It would certainly be good to avoid a physical switch. Very easy to forget to turn it off.
There's a bunch of pins to take care of in turnOn(), turnOff() which put a mixture of levels on the data pins. I made an error when I first built this and had gnd only on pin 1, not on 12 or the end of  the pulldown resistors. Inserting a GAL turned on the LED connected to Vcc - it was getting power through those control pins. We need to make sure everything can float up to Vcc together.

**Comment by ole00** [comment]
turnOn / turnOff() : I've just fixed that and pushed the fix to git. Hopefully there will be less accidents when inserting GAL into the socket now.

**Comment by artgodwin** [comment]
Thanks, I'll use that version when I try this again with changes (resistor in line with pin 1). I may modify the Vpp too - perhaps leave the 3608  connected but isolate Vpp unless Vcc is present.

**Comment by artgodwin** [comment]
I found another programmer, also based on GALbalast and ATFblast. https://www.ythiee.com/2021/06/06/galmate-hardware/
Unfortunately the firmware appears to be missing. I'm not sure if that's intentional, but neither do I see a place to purchase the micro.

edit:
In fact the AVR code is there, in the GAL_Hardware zipfile. I was looking for his determination of pulse time but it seems to work out the same as yours (both derived from GALblast I guess). The only points of note I've seen so far are that Vpp should rise slowly (it does, anyway, for the 3608 .. though maybe I should check if it's given long enough to settle) and that there are some comments about setting SDIN low at a few points as though he got that wrong initially.

The control code is written in 'purebasic' which is a nuisance. It costs $79 for a personal licence. He does provide a windows binary but that's not very useful to me.

**Comment by ole00** [comment]
GALMate: thanks - I was not aware of that project. The author seems very knowledgeable about GALs, and his published findings might be useful.  VPP rising slowly - yes I read that, so far it seems to affect ST brand of GALs if I understood it correctly. I checked the basic source code as well, also found it hard to read (because of the Basic syntax). But if needed, with a bit of effort it can be deciphered. What I'm interested about is how he drives ATF16V8C which is not supported by Afterburner. Good  that you found the MCU source code (thanks to Yorck for publishing it in the first place) - I could not find it either, so I contacted him over email.  The compiled basic code seems to run on Wine (on my Linux desktop), and there is a way how to redirect Wine's serial ports (COM1 etc) to physical serial ports (/dev/ttyUSB0 etc).

Both mt3608 and XL6009 have a pin called SW (switch) which might be the solution. The afterburner's physical switch could either pull the SW pin high, or it could be controlled via Arduino's A0 pin programatically. This needs to be investigated how feasible it is and whether it improves the power rail dip during turn on. I bought the XL6009 module and so far I have better impression of it than mt3608. When the mt3608 is driven above 16V it produces buzzing sound, probably works harder to achieve that voltage. XL6009 seem to be OK deriving 24V from Arduino's 5V rail. Also - as you noted - the pins are easier to access for modding and soldering.

**Comment by artgodwin** [comment]
I will be interesting to see what you can do with it. I don't think pulling it high will work, but disconnecting it with a physical switch should do something similar to enable - stop the boost, leaving it at 5V. However I think it will also disable during power up and so be better behaved than  'enable' . Unfortunately we don't really want that operated by a physical switch and an electronic switch is awkward as it has to operate  at Vpp. Might as well just put a switch in Vpp.

I think tweaking the feedback has some promise, either a complete setting with a digital pot as you've suggested, or a couple of transistors switching alternative values of the 2k2 FB voltage-divider resistor to gnd.

**Comment by ole00** [comment]
I made a prototype of the programmable voltage based on resistor array connected to Arduino's analogue pins and the FB pin on MT3608. When I control 4 standard value resistors (6k8, 4k7, 4k (3k3 + 680R),3k3) I get these VPP values on the MT3608 output:
4.8, 7.8, 9.1, 10.0, 11.0, 12.2, 13.0, 14.1, 14.4, 15.5, 16.2

The MT3608 must still be tuned by the Pot, but the tuning is done only once to set an appropriate wiper/resistance ratio. Because I'm able to set VPP value 4.8V by changing the resistor array connections (actually all are disconnected from ground and A pins are set as Input) I no longer need to do the EN pin mod on MT3608. MT3608 still needs a mod though, A 51k resistor has to be soldered on the bottom of the module, in between the wiper (FB) and one of the other Pot pin. That is an easier mod than the current mod because the POT terminals are bigger than the SMT pads. The resistor array also need additional 10k resistor in parallel leading to ground to achieve the above VPP values.

With this arrangement I can detect presence of the VPP resistor array by reading the Analogue values on A pins and comparing the values with the expected values. Therefore the Arduino code can be common for current Afterburners design and the new one by doing the resistor array test early when the Arduino starts. By reading the analogue values I can even detect whether the Pot tuning is completely off and needs calibration. It is not super precise, but it's close enough to allow some sort of tuning gage as a visual aid
for tuning. The fine tuning must be done with the help of multimeter.

So far I used only 4 pins to achieve the above values. I could probably add another resistor to have a bit more VPP values (like 10.5V, 11.5 etc) if needed. The measured values (by multimeter) vary a bit compared to calculated results, but that's because of the 1% resistor tolerances especially when used in combinations. 

The digital Pot variant is still not off the table, but the fact that most of the cheap ones have 20% variability might be an issue.  Or perhaps not, because the resistance variability could possibly be corrected by tuning the Pot.  Anyway, the resistor array solution does not need an external IC, so it is very convenient  to build quickly without sourcing the extra IC. The advantage of digital Pot is that it saves at least 2 Arduino pins, which could be used for something else (single ZIF socket solution perhaps? :-)

**Reply by artgodwin** [reply]
Nice result !
I hadn't thought of using multiple resistors in parallel - the best combination I could think of was to leave the existing pot in place to set 12V (or 10 - whatever the read voltage is) and add a pot for the high voltage. This still needs 2 pins + the enable (for 5V), I like your scheme much better, especially as you've managed to make it nonlinear and hence have more usable combinations, avoiding too many dead valkes between 5 and 10.  I think an I2C digital pot (or serial DAC) rather than inc/dec or SPI could save more pins. Conveniently, the I2C pins are A4,A5 on a Uno (though it can be bitbanged on other pins).

But I guess the exact voltages aren't that important. Doesn't the GAL give only 0.5V resolution ? Probably anything within 0.5V of the spec. will work OK.

You've listed 11 voltages above but there should be 16 combinations. Are the others close to duplicates ? Or too low to be useful ? 
Also an R-2R ladder should give useful combinations without needing lots of resistor values though I think it loses out by being linear - you end up with a lot of useless voltages. Amy help to use the structure, but not the ratio R/2R.
https://microcontrollerslab.com/r-2r-ladder-dac-digital-to-analog-converter-working-examples-circuits/

I was thinking that having no gnd resistor would set FB = Vout and it would go to 0.6V. But of course it's a boost converter so can't go below Vin. Could use a buck/boost converter instead, perhaps.

Be careful of using analog pins for digital. They work perfectly on a Uno, which only has 5 of them. But some versions of the arduino mini bring out additional pins, often in weird places (on the end of the board, or in the middle). These still work fine as analog inputs but some of the extra pins don't have the digital functionality. This is probably why Arduino chose to include only those that did, so all Arduino Uno / Duemilanove pins can be digital.

**Reply by ole00** [reply]
> You've listed 11 voltages above but there should be 16 combinations. Are the others close to duplicates ? Or too low to be useful ?

The latter. If I simultaneously enable more resistors (connect them to ground) then it divides the final FB resistance (leading to ground) to a small value. Such combinations give few more VPP values, but they are above 16V so not useable. But you have a point - it should be possible to find better resistor combos to employ most of them.  Ideally, I would also like to get rid of the 51k resistor that currently mods the MT3608 FB value and have a direct connection between FB and the resistor array instead. The reason for this is to be able to upgrade already built afterburners with this variable VPP. So, the resistor values would have to be reworked anyway and the values could be optimised to cover the voltage range better. My prototype was just a proof of concept that is can be done this way.

R2R ladder: yes, I'm aware of these. They could be useful, but as you noted they tend to be linear and also require more resistors (space considerations on through hole designs).

Digital POT: another advantage is that it gives more fine-grained resistance control, therefore the VPP could be 'ramped-up' nicely if needed (ST IC's). The resistor array could in theory 'ram-up' as well, but more abruptly.  Also, the ramping up requires to switch / toggle certain resistors and  during a brief time of that switchover certain resistors would be enabled when they should not (basically they are not latched)  - this would cause brief glitches in VPP. It is not an issue if ramping-up is not used because the resistor array state changes only from 'all-off' to 'some-on'. A smoothing cap on VPP could also be attempted to reduce the abruptness of changes. So, technically the digital POT has many advantages (uses fewer pins, fine-grained resistance), but the resistor array is more in the whole spirit of current afterburner design - simplicity.  I'll make a test with the digital POT and then will decide. Voice your opinion as well if you have preference.

Analogue pins - thanks, noted. I'll rely only on A0-A5 as present on UNOs.

**Comment by artgodwin** [comment]
What does setting some pins to +5 do ? It should reduce the output voltage and maybe reduce the step size of other pins. Might be a way to force 5V even though the other resistors are sized for 14V.

**Comment by ole00** [comment]
I have not tried that. 5V should not be an issue, when all A pins are set as input the voltage drops to 4.8V which is (and so far was) good enough for the GAL chip to function normally. That's how the switching between '5V' and 12V (and other) VPP voltages will be done.

**Comment by ole00** [comment]
I made a progress with the new design. The new features include:

- unified socket for 20 pin and 24 devices (including GAL20V8)
- variable VPP via digi pot
- VPP measurement via A0 ADC (used for VPP calibration)
- DPDT switch to disconnect both EDIT (VPP) and 5V rail
- improved the 5V rail dips issue (used 220uF CAP as you suggested)
- single PCB with selectable SMT or TH footprints
- simpler connection to MT3608 module (no trace cuts required)

Extra components compared to the old design:
- MCP4131-103 digi pot (for variable VPP)
- 74HC595 (acts as a GPIO extender)
- MOSfet for 20-pin socket GND
- Trim Pot 2k for compensation of MCP4131 (in)tolerances, can be replaced with fixed resistor
- more passives


The new software (unreleased yet) already works with all planned features (I managed to destroy few GALs during the process..).
It is also backward compatible with the old board design (I can detect the new design by probing the presence of MCP4131 digi pot).
I just need to validate the second revision of the new PCB and then prepare some documentation and release it.

A teaser image: top - the first revision with the 595 SMT footprint badly placed, bottom second revision with better placement.
![afterburner_wip_3 0](https://user-images.githubusercontent.com/13712137/227808302-2b8d51ef-d906-4201-9f05-47f2a522cf2b.jpg)

**Comment by artgodwin** [comment]
Thanks. I'll give it a try when you're happy with it.
I built up the other design but it doesn't work yet. As well as needing a
paid-for version of the compiler to change anything, It has a nasty comms
method - it delineates data frames by forcing a break condition. This
doesn't work well on an arduino platform for a couple of reasons.

**Reply by ole00** [reply]
Ah, you mean GALmate? Its PC software does not seem that complicated- at least for the GAL programming part. It sends a  packet of bytes to Atmega, to write or read fuses at certain row. It should be possible to reimplement the core functionality in C reasonably quickly. I checked it when I was implementing the ATF16V8C support, and after a while it become clear.

**Comment by artgodwin** [comment]
Yes, but the point of building it was to quickly compare another
implementation to rty to understand why I couldn't program 20V8s. I was
trying to avoid re-coding the PC end (and probably adding a few bugs :)

**Comment by artgodwin** [comment]
https://www.techtarget.com/whatis/definition/yak-shaving

**Comment by ole00** [comment]
> I couldn't program 20V8s

I see. The 20V8B's I bought from Aliexpress work OK on both old design (with adapter) and the new design (no adapter). I have one 20V8B that I somehow managed to erase the PES, but that was the only hiccup I had. I noticed GALmate supports rewriting the PES, so I'll try to implement that into Afterburner.  The PES data on those GALs  state to use 14.5VPP, but that usually does not work well with Afterburner, so it reports 2V less.  I programmed it with VPP 12V (old design) and 12.5V new design. I also checked GALep's software which is free to download (I do not have GALep programmer though) which has a human readable definition of GALs it can program. It is a text file containing some configuration data. One of the field is VPP. For ATF16V8C the VPP they use is 11V , so it kind of confirms my finding that 12V for ATF's is too much.  For Lattice 20V8 they use 12V - the same what I used.

**Comment by crystalct** [comment]
Hello, someone want to sell one Afterburner already done?
